<a href="https://colab.research.google.com/github/rennanvra/Praktikum-IPSD/blob/main/MODUL5/guided5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### **Machine Learning**

1. Supervised (memiliki label)
- Regresi (variabel target nilai kontinu, ex: prediksi data umur, harga saham)

  Contoh: Linear Regression

- Klasifikasi (diskrit)

2. Unsupervised (tanpa kelas atau label, hanya mempunyai variabel x)
- Clustering (ex: pengelompokan jeni spelanggan dengan memasukkan pendapatan dan pengeluaran pelanggan)

  Contoh: Kmeans, DBScan, Hierarcical Clustering.

3. Reinforcement (reward)

### **Regresi**


- Regresi adalah konsep awal dari algoritma untuk memprediksi nilai
- Titik-titik pada visualisasi disebut data asli
- Garis pada visualisasi disebut garis prediksi
- Semakin kecil bias atau error semakin bagus nilai atau hasilnya
- Cost function: Menghitung rata-rata error untuk seluruh data train
- Loss Function: Menghitung error untuk satu data point.


1. R^2 : Memeriksa apakah seberapa besar pengaruh antara variabel satu ke variabel lainnya. Semakin mendekati nilai 1 semakin bagus, semakin berpengaruh.
2. MAE (Mean Absolute Error): Menghitung rata-rata selisih absolut antara nilai prediksi dan nilai sebenarnya.
3. MSE (Mean Squared Error) : Menghitung rata-rata kuadrat selisih antara nilai prediksi dan nilai sebenarnya.
4. RMSE (Root Mean Squared Error) : Akar kuadrat dari MSE.


#### **Uji Asumsi Klasik**

Uji asumsi klasik dilakukan untuk memastikan bahwa model regresi linear yang dibangun memenuhi asumsi-asumsi dasar. Jika asumsi tidak terpenuhi, maka hasil analisis regresi mungkin tidak valid.

1. Uji Normalitas : memastikan bahwa residual (error) terdistribusi normal.
2. Uji Linearitas : memastikan bahwa terdapat hubungan linear antara variabel independen dan dependen.
3. Uji Homoskedastisitas : memastikan bahwa varians residual konstan untuk semua nilai variabel independen.
4. Uji Autokorelasi : memastikan bahwa tidak ada korelasi antara residual pada observasi yang berbeda.
5. Uji Multikolinearitas : memastikan bahwa tidak ada korelasi yang tinggi antara variabel independen.


#### **Regularisasi**

Digunakan untuk mencegah overfitting (perbedaan antara data training dan testing, contoh training 98% dan testing 88%). Overfitting terjadi ketika model machine learning terlalu kompleks dan menghafal data training, bukannya mempelajari pola umum. Akibatnya, model bekerja dengan baik pada data training tetapi buruk pada data baru yang belum pernah dilihat sebelumnya.

1. L1 Regularization (Lasso Regression)
- Menambahkan penalty berupa jumlah nilai absolut dari koefisien.
- Dapat menghasilkan model yang sparse, di mana beberapa koefisien menjadi nol.
- Berguna untuk feature selection.

2. L2 Regularization (Ridge Regression)

- Menambahkan penalty berupa jumlah kuadrat dari koefisien.
- Mencegah koefisien menjadi terlalu besar.
- Berguna untuk mengurangi overfitting.

#### **Regresi Polinomial Linear**

Regresi polinomial linear adalah bentuk regresi linear di mana hubungan antara variabel independen dan dependen dimodelkan sebagai polinomial.

In [ ]:
# Import library yang dibutuhkan
from sklearn.datasets import load_diabetes  # Load dataset diabetes
from sklearn.linear_model import LinearRegression, Lasso, Ridge  # Model regresi
from sklearn.model_selection import train_test_split  # Membagi dataset
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score  # Metode evaluasi

In [ ]:
# Load dataset diabetes
x, y = load_diabetes(return_X_y=True)

In [ ]:
# Inisialisasi model regresi linear
lr = LinearRegression() #model

In [ ]:
lr.fit(x, y) # Melatih model dengan data
y_pred = lr.predict(x) # Memprediksi nilai dengan data yang sama

In [ ]:
# Menampilkan nama fitur dan jumlahnya
load_diabetes()['feature_names'] #variabel data

['age', 'sex', 'bmi', 'bp', 's1', 's2', 's3', 's4', 's5', 's6']

In [ ]:
len(load_diabetes()['feature_names']) # Menampilkan panjang fitur data

10

In [ ]:
# Evaluasi model dengan R2, MSE, dan MAE menggunakan semua data (train & test tidak dipisah)
print(r2_score(y, y_pred))  # Mengukur R2 score
print(mean_squared_error(y, y_pred))  # Mengukur Mean Squared Error
print(mean_absolute_error(y, y_pred))  # Mengukur Mean Absolute Error

0.5177484222203499
2859.69634758675
43.27745202531506


### Interpretasi

1. R^2 digunakan untuk memeriksa seberapa besar pengaruh variabel atau nilai yang di prediksi. Semakin mendekati angka 1 maka data semakin berpengaruh atau baik. Nilai di atas memang lebib dari 0,5 sekian namun masih menandakan bahwa data belum cukup optimal untuk. Nilai 0,5177 menunjukkan bahwa model dapat menjelaskan sekitar 51.77% variabilitas dari data target menggunakan fitur yang ada.

2. MSE digunakan untuk menghitung rata-rata kuadrat selisih antara nilai prediksi dan nilai sebenarnya. Nilai MSE sebesar 2859.70 menunjukkan bahwa kesalahan rata-rata kuadrat antara prediksi dan nilai sebenarnya cukup tinggi. Semakin rendah nilai MSE, semakin baik performa model, sehingga nilai ini menunjukkan masih adanya kesalahan prediksi yang cukup besar.

3. MAE digunakan untuk menghitung rata-rata selisih absolut antara nilai prediksi dan nilai sebenarnya. Ini berarti bahwa secara rata-rata, prediksi model meleset sekitar 43.28 unit dari nilai target sebenarnya.

## Teknik pembagian data

1. Holdout Splitting : Hanya membagi dua bagian (training dan testing)
2.


In [ ]:
from sklearn.model_selection import train_test_split

# Memisahkan data menjadi data latih dan uji (80% latih, 20% uji)
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

In [ ]:
# Inisialisasi model regresi linear
lr = LinearRegression()
# Melatih model menggunakan data latih
lr.fit(x_train, y_train)
# Prediksi menggunakan data uji
y_pred = lr.predict(x_test)

# Evaluasi model pada data uji
print(r2_score(y_test, y_pred))  # Mengukur R2 score pada data uji
print(mean_squared_error(y_test, y_pred))  # Mengukur Mean Squared Error pada data uji

0.4526027629719195
2900.193628493482


### Interpretasi

1. R^2 pada data ini jauh lebih rendah, yang berarti menunjukkan bahwa model mungkin sedikit overfitting pada data latih.

2. Nilai MSE pada data uji sebesar 2900.19 mirip dengan MSE pada data latih, yang berarti tingkat kesalahan prediksi model relatif konsisten antara data latih dan data uji. Namun, nilai ini masih menunjukkan kesalahan prediksi yang cukup besar.